src/main/java/uk/ac/ebi/ega/egafuse/service/EgaFileService.java

```diff
@@ -86,19 +103,64 @@ public EgaFileService(OkHttpClient okHttpClient, String apiURL, Token token) {
                     filename = filename.substring(filename.lastIndexOf("/") + 1);
                 }
 
-                String type = "SOURCE";
                 if (filename.toLowerCase().endsWith(".gpg")) {
-                    type = "GPG";
                 } else if (filename.toLowerCase().endsWith(".cip")) {
-                    type = "CIP";
+                    file.setFileSize(file.getFileSize() - 16);
                 }
 
-                egaFiles.add(new EgaFile(filename.substring(0, filename.length() - 4), type, file));
+                egaFiles.add(new EgaFile(filename.substring(0, filename.length() - 4), file, this));
             }
             return egaFiles;
         default:
             LOGGER.error("status: {}", status);
             throw new ClientProtocolException(response.body().string());
         }
     }
+
+    public int fillBufferCurrentPage(Pointer buffer, String fileId, long totalFileSize, long currentPageSize,
+            long offset) {
+        int bytesToRead = (int) Math.min(totalFileSize - offset, currentPageSize);
+        int currentPage = (int) (offset / PAGE_SIZE);
+
+        if (offset >= totalFileSize || bytesToRead <= 0)
+            return -1;
+
+        loadPage(totalFileSize, fileId, currentPage, totalFileSize);
+
+        byte[] page = null;
+        final String key = fileId + "_" + currentPage;
+        try {
+            if (keys.get(key).get() != null) {
+                page = (byte[]) cachemanager.getCache("archive").get(key).get();
```

Comment:
enum/constant can be used e.g. "archive"; also .getCache("archive").get(key) may return null hence call to .get() may fail.

Reply:
Added null checks & archive as a constant variable.